### PR TITLE
more robust intersection with parent

### LINKED ISSUE #151: ocrd-tesserocr-segment-region fails with AttributeError: 'LineString' object has no attribute 'exterior'
Error message:

    19:25:24.366 INFO processor.TesserocrSegmentRegion - INPUT FILE 88 / PHYS_0089
    19:25:24.637 INFO processor.TesserocrSegmentRegion - Page 'PHYS_0089' images will use 400 DPI from image meta-data
    19:25:24.637 INFO processor.TesserocrSegmentRegion - Detecting regions in page 'PHYS_0089'
    19:25:25.936 INFO processor.TesserocrSegmentRegion - Detected region 'region0000': 899,571 2899,575 2898,932 898,928 (FLOWING_TEXT)
    19:25:25.936 INFO processor.TesserocrSegmentRegion - Detected region 'region0001': 1192,943 2672,946 2672,1105 1192,1102 (FLOWING_TEXT)
    19:25:25.936 INFO processor.TesserocrSegmentRegion - Detected region 'region0002': 1631,1180 2246,1182 2246,1266 1631,1264 (FLOWING_TEXT)
    19:25:25.937 INFO processor.TesserocrSegmentRegion - Detected region 'region0003': 884,1333 2990,1337 2985,4141 879,4137 (FLOWING_TEXT)
    Traceback (most recent call last):
      File "/OCR-D/venv-20200912/bin/ocrd-tesserocr-segment-region", line 8, in <module>
        sys.exit(ocrd_tesserocr_segment_region())
      File "/OCR-D/venv-20200912/lib/python3.7/site-packages/click/core.py", line 829, in __call__
        return self.main(*args, **kwargs)
      File "/OCR-D/venv-20200912/lib/python3.7/site-packages/click/core.py", line 782, in main
        rv = self.invoke(ctx)
      File "/OCR-D/venv-20200912/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
        return ctx.invoke(self.callback, **ctx.params)
      File "/OCR-D/venv-20200912/lib/python3.7/site-packages/click/core.py", line 610, in invoke
        return callback(*args, **kwargs)
      File "/OCR-D/venv-20200912/lib/python3.7/site-packages/ocrd_tesserocr/cli.py", line 16, in ocrd_tesserocr_segment_region
        return ocrd_cli_wrap_processor(TesserocrSegmentRegion, *args, **kwargs)
      File "/OCR-D/venv-20200912/lib/python3.7/site-packages/ocrd/decorators.py", line 102, in ocrd_cli_wrap_processor
        run_processor(processorClass, ocrd_tool, mets, workspace=workspace, **kwargs)
      File "/OCR-D/venv-20200912/lib/python3.7/site-packages/ocrd/processor/helpers.py", line 69, in run_processor
        processor.process()
      File "/OCR-D/venv-20200912/lib/python3.7/site-packages/ocrd_tesserocr/segment_region.py", line 172, in process
        self._process_page(layout, page, page_image, page_coords, input_file.pageId)
      File "/OCR-D/venv-20200912/lib/python3.7/site-packages/ocrd_tesserocr/segment_region.py", line 222, in _process_page
        polygon2 = polygon_for_parent(polygon, page)
      File "/OCR-D/venv-20200912/lib/python3.7/site-packages/ocrd_tesserocr/segment_region.py", line 339, in polygon_for_parent
        return interp.exterior.coords[:-1] # keep open
    AttributeError: 'LineString' object has no attribute 'exterior'

### PR DESCRIPTION
**Pull request by bertsky**
Fix #151

### CODE REVIEW

**Comment by bertsky** [comment]
Fix #149

**Comment by stweil** [comment]
A test with the first commit and urn:nbn:de:bsz:180-digad-35210 passed without failures.

File: ocrd_tesserocr/segment_region.py [hunk 1]
```diff
@@ -326,14 +327,39 @@ def polygon_for_parent(polygon, parent):
...
+        # heterogeneous result: filter zero-area shapes (LineString, Point)
+        interp = unary_union([geom for geom in interp.geoms if geom.area > 0])
     if interp.type == 'MultiPolygon':
+        # homogeneous result: construct convex hull to connect
+        # FIXME: construct concave hull / alpha shape
         interp = interp.convex_hull
+    if interp.minimum_clearance < 1.0:
+        # follow-up calculations will necessarily be integer;
+        # so anticipate rounding here and then ensure validity
+        interp = asPolygon(np.round(interp.exterior.coords))
```
Comment (stweil) [inline ocrd_tesserocr/segment_region.py:360]: The code fails here now: `NameError: name 'np' is not defined`

**Review by stweil — APPROVED** [review]
My test case now passes (mostly). Therefore I suggest to merge this.

**Comment by stweil** [comment]
@kba (or whoever has merge rights), can this PR be merged? I'd like to have a working `ocrd_all` again, and this seems to be one required component.

**Comment by bertsky** [comment]
> @kba (or whoever has merge rights), can this PR be merged? I'd like to have a working `ocrd_all` again, and this seems to be one required component.

#153 is really the same thing for the other processors, so we should first incorporate the same logic there. It will be more consistent here (with 0.9.3 to fix _all_ currently known segmentation bugs), but also in ocrd_all (all affected modules joining in one commit, no extra delay for the next patch wave).

**Review by kba — APPROVED** [review]
@bertsky you have the best overview on how the PR work together, merge at will :)

**Review by stweil — APPROVED** [review]
Please merge.

**Pull request merged**